taichi/ir/frontend_ir.cpp

```diff
@@ -970,10 +991,25 @@ Expr ASTBuilder::expr_alloca() {
   return var;
 }
 
+Expr make_local_matrix(const std::vector<int> &shape,
+                       const DataType &dt,
+                       const std::vector<Expr> &elements) {
+  return Expr(std::make_shared<MatrixExpression>(elements, shape, dt));
+}
+
 Expr ASTBuilder::expr_alloca_local_tensor(const std::vector<int> &shape,
                                           const DataType &element_type,
                                           const ExprGroup &elements,
                                           std::string tb) {
+  if (this->use_real_matrix_) {
+    auto matrix_expr = make_local_matrix(shape, element_type, elements.exprs);
+    auto v = this->expr_alloca();
+    this->expr_assign(v, matrix_expr, tb);
+    // type check for variable `v` since
+    // expr_assign couldn't propagate the info
+    v->ret_type = matrix_expr.cast<MatrixExpression>()->dt;
```

Comment:
I guess the constructor of `FrontendAssignStmt` called in `expr_assign` propagates the info?